lib/ecto/type.ex

```diff
@@ -916,92 +932,141 @@ defmodule Ecto.Type do
-       when type in [:naive_datetime, :naive_datetime_usec] do
-    NaiveDateTime.compare(term1, term2) == :eq
+
+  defp equal_fun(mod) when is_atom(mod) do
+    if loaded_and_exported?(mod, :equal?, 2) do
+      &mod.equal?/2
+    end
   end
-  defp do_equal?(type, %DateTime{} = term1, %DateTime{} = term2)
-       when type in [:utc_datetime, :utc_datetime_usec] do
-    DateTime.compare(term1, term2) == :eq
+
+  defp equal_decimal?(%Decimal{} = a, %Decimal{} = b), do: Decimal.equal?(a, b)
+  defp equal_decimal?(_, _), do: false
+
+  defp equal_time?(%Time{} = a, %Time{} = b), do: Time.compare(a, b) == :eq
+  defp equal_time?(_, _), do: false
+
+  defp equal_utc_datetime?(%DateTime{} = a, %DateTime{} = b), do: DateTime.compare(a, b) == :eq
+  defp equal_utc_datetime?(_, _), do: false
+
+  defp equal_naive_datetime?(%NaiveDateTime{} = a, %NaiveDateTime{} = b),
+    do: NaiveDateTime.compare(a, b) == :eq
+  defp equal_naive_datetime?(_, _),
+    do: false
+
+  defp equal_list?(fun, [x | xs], [y | ys]), do: fun.(x, y) and equal_list?(fun, xs, ys)
+  defp equal_list?(_fun, [], []), do: true
+  defp equal_list?(_fun, _, _), do: false
+
+  defp equal_map?(_fun, map1, map2) when map_size(map1) != map_size(map2) do
+    false
   end
-  defp do_equal?(type, term1, term2) when type in @base do
-    term1 == term2
+
+  defp equal_map?(fun, %{} = map1, %{} = map2) do
+    equal_map?(fun, Map.to_list(map1), map2)
   end
-  defp do_equal?(module, term1, term2) when is_atom(module) do
-    module.equal?(term1, term2)
+
+  defp equal_map?(fun, [{key, val} | tail], other_map) do
+    case Map.fetch(other_map, key) do
+      {:ok, other_val} ->
+        fun.(val, other_val) and equal_map?(fun, tail, other_map)
+
+      :error ->
+        false
+    end
   end
-  defp do_equal?({:array, type}, xs, ys) do
-    equal_list?(type, xs, ys)
+
+  defp equal_map?(_fun, [], _) do
+    true
   end
-  defp do_equal?({:map, _}, map1, map2) when map_size(map1) != map_size(map2) do
+
+  defp equal_map?(_fun, _, _) do
     false
   end
-  defp do_equal?({:map, type}, map1, map2) do
-    equal_map?(type, Map.to_list(map1), Map.to_list(map2))
+
+  ## Helpers
+
+  # Checks if a value is of the given primitive type.
+  defp of_base_type?(:any, _), do: true
+  defp of_base_type?(:id, term), do: is_integer(term)
+  defp of_base_type?(:float, term), do: is_float(term)
+  defp of_base_type?(:integer, term), do: is_integer(term)
+  defp of_base_type?(:boolean, term), do: is_boolean(term)
+  defp of_base_type?(:binary, term), do: is_binary(term)
+  defp of_base_type?(:string, term), do: is_binary(term)
+  defp of_base_type?(:map, term), do: is_map(term) and not Map.has_key?(term, :__struct__)
+  defp of_base_type?(:decimal, value), do: Kernel.match?(%Decimal{}, value)
+  defp of_base_type?(:date, value), do: Kernel.match?(%Date{}, value)
+  defp of_base_type?(_, _), do: false
+
+  # TODO: pasted from array/4, remove after changing dump/1 and load/1
```

Comment:
Is this TODO still relevant? :)

Reply:
oops, nope; we need both array/3 and array/4